File: Week 1/cpp/3.better_hello_world.cpp
/*
 * WEEK 1 - C++ FUNDAMENTALS
 * Topic: Output without/with newline (cout, '\n', std::endl)
 * File: 3.better_hello_world.cpp
 *
 * CONCEPT:
 * C++ has no pair `print` / `println`. Instead, the newline is just another
 * value you stream out. `'\n'` is a fast plain newline; `std::endl` is a
 * newline + flush of the output buffer.
 *
 * KEY POINTS:
 *  - `std::cout << "text";`               -> no newline (Java's print)
 *  - `std::cout << "text" << '\\n';`        -> newline (Java's println)
 *  - `std::cout << "text" << std::endl;`   -> newline + flush
 *  - Use `'\\n'` inside loops; reserve `std::endl` for when you need a flush.
 *  - Multiple values can be chained with `<<` operators.
 *
 * SYNTAX:
 *   cout << "abc" << "def" << '\n';
 *   cout << "no newline";
 *
 * DRY RUN:
 *  Three "Hello World" lines, then three on the same line, then a final newline.
 */

#include <iostream>

int main() {
    // println-style: three separate lines
    std::cout << "Hello World" << '\n';
    std::cout << "Hello World" << '\n';
    std::cout << "Hello World" << '\n';

    // print-style (no newline): all on the same line
    std::cout << "Hello World";
    std::cout << "Hello World";
    std::cout << "Hello World";
    std::cout << '\n';

    // std::endl flushes — useful before reading interactive input
    std::cout << "Flushed line" << std::endl;

    return 0;
}

/*
 * NOTES:
 *  - Java: `print` / `println`     | C++: stream `'\n'` or `std::endl`.
 *  - Python: `end=` parameter      | C++: include or omit `'\n'`.
 *  - `std::endl` = `'\n'` + `flush`; in tight loops prefer `'\n'`.
 *  - For "printf-style" formatting use `<cstdio>`:
 *      printf("x = %d\n", 42);
 *    or `<format>` (C++20):
 *      std::cout << std::format("x = {}\n", 42);
 */

File: Week 1/python/3.better_hello_world.py
"""
WEEK 1 - PYTHON FUNDAMENTALS
Topic: print() — newlines, separators, and the `end` parameter
File: 3.better_hello_world.py

CONCEPT:
Java offers `System.out.print()` (no newline) and `System.out.println()`
(with newline). Python has a SINGLE function — `print()` — and you control
the trailing character with the keyword argument `end`. By default,
`end="\\n"`, which mimics Java's `println`. Setting `end=""` mimics `print`.

KEY POINTS:
- `print(*objects, sep=' ', end='\\n', file=sys.stdout, flush=False)`
- Multiple println-equivalent calls each end with their own newline
- Setting `end=""` keeps the next print on the same line
- `sep` controls how multiple positional args are joined
- `flush=True` forces the output buffer to flush immediately

SYNTAX:
print("text")             # prints "text\n"  (Java: System.out.println)
print("text", end="")     # prints "text"    (Java: System.out.print)
print("a", "b", "c")       # prints "a b c\n"
print("a", "b", sep="")    # prints "ab\n"

DRY RUN:
1. Three calls to `print("Hello World")` -> three lines on stdout.
2. Three calls to `print("Hello World", end="")` -> single line:
       Hello WorldHello WorldHello World
3. `print("X", "Y", sep="--", end="!\\n")` -> "X--Y!" then newline.
"""


def main() -> None:
    print("--- println-style (default end='\\n') ---")
    print("Hello World")
    print("Hello World")
    print("Hello World")

    print("\n--- print-style (end='') ---")
    print("Hello World", end="")
    print("Hello World", end="")
    print("Hello World")  # last one gets a newline

    print("\n--- mixed end values ---")
    print("Same", end=" ")
    print("line", end=" ")
    print("here.")

    print("\n--- custom separator with multiple args ---")
    print("X", "Y", "Z")               # default: "X Y Z"
    print("X", "Y", "Z", sep="")        # "XYZ"
    print("X", "Y", "Z", sep="--")      # "X--Y--Z"
    print("X", "Y", "Z", sep="--", end="!\n")  # "X--Y--Z!"


if __name__ == "__main__":
    main()


# NOTES:
# - Java's `print` vs `println` is a method choice; Python's is a parameter (`end=`).
# - Python's `print` joins multiple args with a space automatically (Java would
#   require explicit string concatenation with the `+` operator).
# - For formatted output, Python prefers f-strings: print(f"x = {x}").
#   Java's equivalent would be `System.out.printf("x = %d%n", x)`.
